File: test/pvo_vtu_file.c

/// vim: tabstop=4:expandtab:hlsearch

#include "../pvo.h"
#include "../pvo_report.h"
#include "../pvo_memory.h"

#include <iso646.h>

int main( int argc, char** argv ) {
    pvo_cookie_t cookie;
    pvo_vtu_file_t fh;
    pvo_float3_t pt[8];
    int32_t cia[2] = { 0, 8 };
    int32_t cja[8] = { 0, 1, 2, 3, 4, 5, 6, 7 };
    int   X[] = { 2, 2, 3, 0, -1, 3, 5, -7 };
    float Y[1];

    MPI_Init( &argc, &argv );
    pvo_init( MPI_COMM_WORLD );

    pt[0][0] = 2*pvo_world_rank() + 0.5;
    pt[0][1] = 2*pvo_world_rank() + 0.5;
    pt[0][2] = 2*pvo_world_rank() + 0.5;

    pt[1][0] = 2*pvo_world_rank() + 1.5;
    pt[1][1] = 2*pvo_world_rank() + 0.5;
    pt[1][2] = 2*pvo_world_rank() + 0.5;

    pt[2][0] = 2*pvo_world_rank() + 1.5;
    pt[2][1] = 2*pvo_world_rank() + 1.5;
    pt[2][2] = 2*pvo_world_rank() + 0.5;

    pt[3][0] = 2*pvo_world_rank() + 0.5;
    pt[3][1] = 2*pvo_world_rank() + 1.5;
    pt[3][2] = 2*pvo_world_rank() + 0.5;

    pt[4][0] = 2*pvo_world_rank() + 0.5;
    pt[4][1] = 2*pvo_world_rank() + 0.5;
    pt[4][2] = 2*pvo_world_rank() + 1.5;

    pt[5][0] = 2*pvo_world_rank() + 1.5;
    pt[5][1] = 2*pvo_world_rank() + 0.5;
    pt[5][2] = 2*pvo_world_rank() + 1.5;

    pt[6][0] = 2*pvo_world_rank() + 1.5;
    pt[6][1] = 2*pvo_world_rank() + 1.5;
    pt[6][2] = 2*pvo_world_rank() + 1.5;

    pt[7][0] = 2*pvo_world_rank() + 0.5;
    pt[7][1] = 2*pvo_world_rank() + 1.5;
    pt[7][2] = 2*pvo_world_rank() + 1.5;

    Y[0] = 2.3*pvo_world_rank();

    if( -1 == pvo_cookie_create( pvo_world_rank()%2, &cookie ))
        PVO_DIE( "pvo_cookie_create() failed." );

    if( -1 == pvo_cookie_insert_var( cookie, PVO_VAR_NODEDATA, PVO_VAR_INT32  , 8, "X", X ))
        PVO_DIE( "pvo_cookie_insert_var() failed." );
    if( -1 == pvo_cookie_insert_var( cookie, PVO_VAR_CELLDATA, PVO_VAR_FLOAT32, 1, "Y", Y ))
        PVO_DIE( "pvo_cookie_insert_var() failed." );

    if( -1 == pvo_vtu_file_open( "test3", cookie, 8, pt, 1, cia, cja, &fh ))
        PVO_DIE( "pvo_vtu_file_open() failed." );

    pvo_file_write( (pvo_file_t )fh );

    if( -1 == pvo_vtu_file_close( fh ))
        PVO_DIE( "pvo_file_vtu_close() failed." );

    if( -1 == pvo_cookie_delete( cookie ))
        PVO_DIE( "pvo_cookie_delete() failed." );

    pvo_quit();
    MPI_Finalize();

    return 0;
}

File: test/pvo_vtu_file.f90

! vi: tabstop=4:expandtab:hlsearch

program pvo_vtu_file_f
    use mpi
    use pvo
    implicit none

    integer :: ierr, num
    integer*8 :: nnodes, ncells
    integer(kind = PVO_COOKIE_KIND) :: cookie
    integer(kind = PVO_VTU_FILE_KIND) :: fh
    real, dimension(3,9) :: pt
    integer, dimension(3) :: cia
    integer, dimension(9) :: cja
    integer, dimension(9) :: X
    real, dimension(2) :: Y
    integer, dimension(2) :: types
   
    call mpi_init( ierr )
    call pvo_init( MPI_COMM_WORLD, ierr )


    cia = (/ 0, 8, 9 /)
    cja = (/ 0, 1, 2, 3, 4, 5, 6, 7, 8 /)
    X   = (/ 2, 2, 3, 0, -1, 2, 5, -7, 8 /)
    
    pt(1,1) = 2*pvo_world_rank() + 0.5
    pt(2,1) = 2*pvo_world_rank() + 0.5
    pt(3,1) = 2*pvo_world_rank() + 0.5

    pt(1,2) = 2*pvo_world_rank() + 1.5
    pt(2,2) = 2*pvo_world_rank() + 0.5
    pt(3,2) = 2*pvo_world_rank() + 0.5

    pt(1,3) = 2*pvo_world_rank() + 1.5
    pt(2,3) = 2*pvo_world_rank() + 1.5
    pt(3,3) = 2*pvo_world_rank() + 0.5

    pt(1,4) = 2*pvo_world_rank() + 0.5
    pt(2,4) = 2*pvo_world_rank() + 1.5
    pt(3,4) = 2*pvo_world_rank() + 0.5

    pt(1,5) = 2*pvo_world_rank() + 0.5
    pt(2,5) = 2*pvo_world_rank() + 0.5
    pt(3,5) = 2*pvo_world_rank() + 1.5

    pt(1,6) = 2*pvo_world_rank() + 1.5
    pt(2,6) = 2*pvo_world_rank() + 0.5
    pt(3,6) = 2*pvo_world_rank() + 1.5

    pt(1,7) = 2*pvo_world_rank() + 1.5
    pt(2,7) = 2*pvo_world_rank() + 1.5
    pt(3,7) = 2*pvo_world_rank() + 1.5

    pt(1,8) = 2*pvo_world_rank() + 0.5
    pt(2,8) = 2*pvo_world_rank() + 1.5
    pt(3,8) = 2*pvo_world_rank() + 1.5

    pt(1,9) = 2*pvo_world_rank() + 1.75
    pt(2,9) = 2*pvo_world_rank() + 1.75
    pt(3,9) = 2*pvo_world_rank() + 1.75

    Y = (/ 2.3*pvo_world_rank(), -1.3*pvo_world_rank() /)

    types = (/ PVO_VTU_HEXAHEDRON, PVO_VTU_VERTEX /)


    call pvo_cookie_create( mod(pvo_world_rank(),2), cookie, ierr )
    if( 0 .ne. ierr ) then
        write(*,*)  'pvo_cookie_create failed'
        call mpi_abort( MPI_COMM_WORLD, 1, ierr )
    endif

    call pvo_cookie_insert_var_int32( cookie, PVO_VAR_NODEDATA, 1, "X"//char(0), X, ierr )
    if( 0 .ne. ierr ) then
        write(*,*)  'pvo_cookie_insert_var_int32 failed'
        call mpi_abort( MPI_COMM_WORLD, 1, ierr )
    endif
    
    call pvo_cookie_insert_var_float32( cookie, PVO_VAR_CELLDATA, 1, "Y"//char(0), Y, ierr )
    if( 0 .ne. ierr ) then
        write(*,*)  'pvo_cookie_insert_var_float32 failed'
        call mpi_abort( MPI_COMM_WORLD, 1, ierr )
    endif

    nnodes = 9
    ncells = 2
    call pvo_vtu_file_open( "test11"//char(0), cookie, nnodes, pt, ncells, cia, cja, types, fh, ierr )
    if(0 .ne. ierr) then
        write(*,*) 'pvo_vtu_file_open failed'
        call mpi_abort( MPI_COMM_WORLD, 1, ierr )
    endif

    call pvo_file_write( fh, ierr )
    if(0 .ne. ierr) then
        write(*,*) 'pvo_file_write failed'
        call mpi_abort( MPI_COMM_WORLD, 1, ierr )
    endif

    call pvo_vtu_file_close( fh, ierr )
    if(0 .ne. ierr) then
        write(*,*) 'pvo_vtu_file_close failed'
        call mpi_abort( MPI_COMM_WORLD, 1, ierr )
    endif

    call pvo_cookie_delete( cookie, ierr )
    if( 0 .ne. ierr ) then
        write(*,*) 'pvo_cookie_delete failed'
        call mpi_abort( MPI_COMM_WORLD, 1, ierr )
    endif
    

    call pvo_quit( ierr )
    call mpi_finalize( ierr )

end program

